building-epics/cross-compile-epics-and-a-ioc-to-an-old-x86-linux.md

```diff
@@ -0,0 +1,367 @@
+
+1.  Create a Virtual Machine and install the target’s Linux distribution. From the Virtual Machine, compile EPICS and my IOC, and then run the IOC in the target
+2.  Build a toolchain configured for my target and use that toolchain to compile both EPICS and the IOC.
+
+The first approach is the easiest, but compiling from inside a VM can be slow and the distribution was not very user friendly. So I took the second path, which I’ll describe in this document.
+
+## Overview
+
+I’m assuming you, like me, are running Ubuntu 64 bits. I’m also assuming you already have EPICS base downloaded and compiled. We will go through the steps of downloading, configuring and compiling [Crosstool-NG](http://crosstool-ng.org/), which will be used to generate our toolchain. Then we will download and compile a couple of libraries needed by EPICS (namely [readline](https://www.gnu.org/software/readline/) and [ncurses](https://www.gnu.org/software/ncurses/)). Finally, we will compile EPICS base and an example IOC to our target using the newly built toolchain.
+
+## Crosstool-NG
+
+### Downloading and extracting
+
+First we get the tarball containing the source code and extract it.
+
+```
+wget http://crosstool-ng.org/download/crosstool-ng/crosstool-ng-1.9.3.tar.bz2
+tar -xvf crosstool-ng-1.9.3.tar.bz2
+```
+
+Crosstool-NG has a lot of dependencies, you might want to get them before compiling:
+
+sudo apt-get install gawk bison flex texinfo automake libtool cvs libcurses5-dev build-essential
+
+### Compiling
+
+In order to compile:
+
+```
+cd crosstool-ng-1.9.3
+./configure --local
+make
+```
+
+### Configuring
+
+Before configuring Crosstool-NG, I gathered information about my target system:
+
+Kernel Version:
+
+```
+$uname -r
+2.6.27.27
+```
+
+GCC Version:
+
+```
+$gcc --version
+gcc (GCC) 4.2.4
+```
+
+glibc version:
+
+```
+$/lib/libc.so.6
+GNU C Library stable release version 2.7, by Roland McGrath et al.
+```
+
+binutils version:
+
+```
+$ld --version
+GNU ld (Linux/GNU Binutils) 2.18.50.0.9.20080822
+```
+
+Based on this information, and on a lot of trial and error, I configured Crosstool-NG as follows (everything else set as default):
+
+```
+./ct-ng menuconfig
+
+PATHS AND MISC OPTIONS
+   \[\*\] Use obsolete features
+   
+TARGET OPTIONS
+   Target architecture (x86)
+   Bitness: (32-bit)
+   (i686) Architecture Level
+   
+OPERATING SYSTEM
+   Target OS (linux)
+   Linux kernel version (2.6.27.55)
+   
+BINARY UTILITIES
+   Binutils version (2.17)
+   
+C-COMPILER
+   GCC version (4.2.4)
+   \[\*\] C++
+   
+C-LIBRARY
+   glibc version (2.6.1)
+   
+```
+
+I tried other configurations, but they crashed the compilation process.
+
+### Compiling the toolchain
+
+Now we can compile the toolchain:
+
+```
+./ct-ng build
+```
+
+This will take a while. Go get some coffee or watch a cat video on Youtube.
+
+Once built, the toolchain will be in $HOME/x-tools/i686-unknown-linux-gnu/
```

Comment:
I would add the $HOME/x-tools/i686-unknown-linux-gnu/ into code blocks as well (single quote)